modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C

```diff
@@ -0,0 +1,114 @@
+      RankTwoScalarTools::scalarOptions(),
+      "Scalar quantity to be computed from tensor and used as a failure criterion");
+  params.addRequiredParam<std::string>("tensor", "The material tensor name.");
+  params.addRequiredCoupledVar("threshold", "The threshold for crack growth.");
+  params.addRequiredParam<bool>(
+      "average", "Should the tensor quantity be averaged over the quadrature points?");
+  params.addParam<Point>(
+      "point1",
+      Point(0, 0, 0),
+      "Start point for axis used to calculate some cylindrical material tensor quantities");
+  params.addParam<Point>(
+      "point2",
+      Point(0, 1, 0),
+      "End point for axis used to calculate some cylindrical material tensor quantities");
+  return params;
+}
+
+XFEMRankTwoTensorMarkerUserObject::XFEMRankTwoTensorMarkerUserObject(
+    const InputParameters & parameters)
+  : XFEMMarkerUserObject(parameters),
+    _tensor(getMaterialProperty<RankTwoTensor>(getParam<std::string>("tensor"))),
+    _scalar_type(getParam<MooseEnum>("scalar_type")),
+    _point1(parameters.get<Point>("point1")),
+    _point2(parameters.get<Point>("point2")),
+    _threshold(coupledValue("threshold")),
+    _average(getParam<bool>("average")),
+    _JxW(_assembly.JxW()),
+    _coord(_assembly.coordTransformation())
+{
+}
+
+bool
+XFEMRankTwoTensorMarkerUserObject::doesElementCrack(RealVectorValue & direction)
+{
+  bool does_it_crack = false;
+  unsigned int numqp = _qrule->n_points();
+
+  if (_average)
+  {
+    Real average_threshold;
+    RankTwoTensor average_tensor;
+    Point average_point;
+    for (unsigned int qp = 0; qp < numqp; ++qp)
+    {
+      if (_threshold[qp] <= 0.0)
+        mooseError("Threshold must be strictly positive in XFEMRankTwoTensorMarkerUserObject");
+      average_threshold += _JxW[qp] * _coord[qp] * _threshold[qp];
+      average_tensor += _JxW[qp] * _coord[qp] * _tensor[qp];
+      average_point += _JxW[qp] * _coord[qp] * _q_point[qp];
+    }
+    Point point_dir;
+    Real tensor_quantity = RankTwoScalarTools::getQuantity(
+        average_tensor, _scalar_type, _point1, _point2, average_point, point_dir);
+    if (point_dir(0) == 0.0 && point_dir(1) == 0.0 && point_dir(2) == 0.0)
+      mooseError("Direction has zero length in XFEMRankTwoTensorMarkerUserObject");
+    direction = point_dir;
+    if (tensor_quantity > average_threshold)
+      does_it_crack = true;
+  }
+  else
+  {
+    unsigned int max_index = std::numeric_limits<unsigned int>::max();
+    Real max_ratio = 0.0;
+    std::vector<Point> directions(numqp);
+    for (unsigned int qp = 0; qp < numqp; ++qp)
+    {
+      if (_threshold[qp] <= 0.0)
+        mooseError("Threshold must be strictly positive in XFEMRankTwoTensorMarkerUserObject");
+      const Real tensor_quantity = RankTwoScalarTools::getQuantity(
+          _tensor[qp], _scalar_type, _point1, _point2, _q_point[qp], directions[qp]);
+      if (directions[qp](0) == 0.0 && directions[qp](1) == 0.0 && directions[qp](2) == 0.0)
```

Comment:
Sure.  I'll do that.  These really will be set to zero in this case, but I'll make it fuzzy.